src/model/channel/message.rs

```diff
@@ -151,6 +151,65 @@ pub struct Message {
     pub poll: Option<Box<Poll>>,
 }
 
+// Custom deserialize function to deserialize components safely without knocking the whole message
+// out when new components are found but not supported.
+fn deserialize_components<'de, D>(deserializer: D) -> Result<Vec<ActionRow>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    #[derive(Deserialize)]
+    struct MinComponent {
+        #[serde(rename = "type")]
+        kind: u8,
+    }
+
+    struct ComponentsVisitor;
+
+    impl<'de> Visitor<'de> for ComponentsVisitor {
+        type Value = Vec<ActionRow>;
+
+        fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            formatter.write_str("a sequence of ActionRow elements")
+        }
+
+        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+        where
+            A: serde::de::SeqAccess<'de>,
+        {
+            let mut components = Vec::new();
+
+            while let Some(raw) = seq.next_element::<&serde_json::value::RawValue>()? {
+                // We deserialize only the `kind` field to determine the component type.
+                // We later use this to check if its a supported component before deserializing the
+                // entire payload.
+                let min_component =
+                    MinComponent::deserialize(raw).map_err(serde::de::Error::custom)?;
+
+                // Action rows are the only top level component supported in serenity at this time.
+                if min_component.kind == 1 {
+                    match ActionRow::deserialize(raw) {
+                        Ok(valid_row) => components.push(valid_row),
+                        Err(_) => {
+                            tracing::debug!("Failed to deserialize ActionRow, malformed data.");
```

Comment:
Should this be ignoring errors?

Reply:
Do you expect top-level action rows to occasionally fail to deserialize once Components V2 are introduced? If not, then the data should be parsed the same as before. Something like:

```rust
components.push(ActionRow::deserialize(raw).map_err(serde::de::Error::custom)?);
```

Reply:
> Do you expect top-level action rows to occasionally fail to deserialize once Components V2 are introduced?

They shouldn't, they haven't changed anything about ActionRows and the allowed component types will remain the same, but I'm not sure if this assumption will change.

The way it is currently should be safer but apply the suggestion is fine by me if you want it to.

Reply:
Given that the layout hasn't changed and Components V2 is shipping soon, they likely won't break it any more in the near future. I'd say the suggestion is fine to apply.

Reply:
Applied in 775a7f30e3a630a591a955dc9c92fec89d2029ed